server/athenian/api/controllers/search_controller/search_prs.py

```diff
@@ -0,0 +1,202 @@
+        tasks["jira_conf"] = get_jira_installation(
+            account_info.account, request.sdb, request.mdb, request.cache,
+        )
+
+    if tasks:
+        results = await gather(*tasks.values())
+        for task_name, result in zip(tasks, results):
+            match task_name:
+                case "repositories":
+                    kwargs["repositories"] = {r.unprefixed for r in result[0]}
+                case "participants":
+                    kwargs["participants"] = result[0] if result else {}
+                case "jira_conf":
+                    kwargs["jira"] = JIRAFilter.from_web(search_req.jira, result)
+
+    return _SearchPRsFilter(**kwargs)
+
+
+@dataclasses.dataclass
+class _SearchPRsFilter:
+    time_from: datetime
+    time_to: datetime
+    repositories: Optional[set[str]] = None
+    participants: Optional[PRParticipants] = None
+    jira: Optional[JIRAFilter] = None
+
+
+@dataclasses.dataclass
+class _SearchPRsAccountInfo:
+    account: int
+    meta_ids: tuple[int, ...]
+    prefixer: Prefixer
+    bots: set[str]
+    release_settings: ReleaseSettings
+    logical_settings: LogicalRepositorySettings
+
+
+@dataclasses.dataclass
+class _SearchPRsConnectors:
+    mdb: DatabaseLike
+    pdb: DatabaseLike
+    rdb: DatabaseLike
+    cache: Optional[aiomcache.Client]
+
+
+async def _search_pr_digests(
+    search_filter: _SearchPRsFilter,
+    account_info: _SearchPRsAccountInfo,
+    connectors: _SearchPRsConnectors,
+) -> list[PullRequestDigest]:
+    mdb, pdb, rdb, cache = connectors.mdb, connectors.pdb, connectors.rdb, connectors.cache
+    if search_filter.repositories is None:
+        repos = set(account_info.release_settings.native.keys())
+    else:
+        repos = search_filter.repositories
+
+    calc = PRFactsCalculator(
+        account_info.account, account_info.meta_ids, mdb, pdb, rdb, cache=cache,
+    )
+    pr_facts = await calc(
+        search_filter.time_from,
+        search_filter.time_to,
+        repos,
+        search_filter.participants or {},
+        LabelFilter.empty(),
+        search_filter.jira or JIRAFilter.empty(),
+        exclude_inactive=True,
+        bots=account_info.bots,
+        release_settings=account_info.release_settings,
+        logical_settings=account_info.logical_settings,
+        prefixer=account_info.prefixer,
+        fresh=False,
+        with_jira=JIRAEntityToFetch.NOTHING,
+    )
+
+    prs_numbers = await fetch_prs_numbers(
+        pr_facts[PullRequestFacts.f.node_id].values, account_info.meta_ids, mdb,
+    )
+    pr_digests = [
+        PullRequestDigest(number=number, repository=repository)
+        for node_id, repository_full_name, number in zip(
+            pr_facts[PullRequestFacts.f.node_id].values,
+            pr_facts[PullRequestFacts.f.repository_full_name].values,
+            prs_numbers,
+        )
+        if number != 0
+        and (repository := account_info.prefixer.prefix_logical_repo(repository_full_name))
```

Comment:
I am worried that we will do this for each pull request. We'd better optimize like this:
```
repo_order = np.argsort(pr_facts[PullRequestFacts.f.repository_full_name].values)
unique_repos, repo_lengths = np.unique(pr_facts[PullRequestFacts.f.repository_full_name].values[repo_order], return_counts=True)
unique_repos = [prefix_logical_repo(r) for r in unique_repos]
prefixed_repos = np.repeat(unique_repos, repo_lengths)[repo_order]  # the order will match pr_facts
```

Reply:
Right, this could become slow with many logical repos.
Given that we know what repos we put in, can we also build the mapping outside the loop ? This is faster, and also avoids complexity in this higher level code:
```
repo_mapping = {r: prefix_logical_repo(r) for r in repos}
PullRequestDigest(number=number, repository=repo_mapping[repository_full_name])
```
This is 5.5 ms against 6.5ms of current loop (with no logical repos).

Reply:
ok good